cephfs/file.go

```diff
@@ -0,0 +1,172 @@
+package cephfs
+
+/*
+#cgo LDFLAGS: -lcephfs
+#cgo CPPFLAGS: -D_FILE_OFFSET_BITS=64
+#include <stdlib.h>
+#include <cephfs/libcephfs.h>
+*/
+import "C"
+
+import (
+	"io"
+	"unsafe"
+)
+
+const (
+	// SeekSet is used with Seek to set the absolute position in the file.
+	SeekSet = int(C.SEEK_SET)
+	// SeekCur is used with Seek to position the file relative to the current
+	// position.
+	SeekCur = int(C.SEEK_CUR)
+	// SeekEnd is used with Seek to position the file relative to the end.
+	SeekEnd = int(C.SEEK_END)
+)
+
+// File represents an open file descriptor in cephfs.
+type File struct {
+	mount *MountInfo
+	fd    C.int
+}
+
+// Open a file at the given path. The flags are the same os flags as
+// a local open call. Mode is the same mode bits as a local open call.
+//
+// Implements:
+//  int ceph_open(struct ceph_mount_info *cmount, const char *path, int flags, mode_t mode);
+func (mount *MountInfo) Open(path string, flags int, mode uint32) (*File, error) {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+	ret := C.ceph_open(mount.mount, cPath, C.int(flags), C.mode_t(mode))
+	if ret < 0 {
+		return nil, getError(ret)
+	}
+	return &File{mount: mount, fd: ret}, nil
+}
+
+func (f *File) validate() error {
```

Comment:
Is this following a standard pattern, or why not a `func (f *File) isValid() bool`?

Reply:
It's the pattern similar to that established in rbd for images. I'd also argue that its preferable to bools in general as bools are low-information versus errors.
Let me know some time if you'd like to hear one of my rants against bools. ;-)

Reply:
But information wise you _are_ returning a bool. You just wrap it in a error/nil pattern. I would understand, if the additional information in the err is used to print the error string in a general context later, but you are not doing that. For me that looks like code bloat. :thinking:

Reply:
Right now yeah. But everywhere we call it would then be followed by a `return ..., ErrNotConnected` which seems less DRY to me. This may end up being an agree-to-disagree situation :-)

Reply:
Did you see my related comment from line 63? I'm happy to agree-to-disagree as soon as I'm sure we don't misunderstand each other. ;-) Where would you use return? So far I only see it would replace the (what I personally find quite ugly) go-typical `if err := f.validate(); err != nil` with a _dry_ `if !f.isValid()`, which is a lot easier to my eyes. But that's my last attempt, promised. If you still prefer that, I will just shrug in wonder and let you go ahead. :-)

Reply:
Yes, I don't see it as an improvement though. When the reference pointer to mount is nil I want to return an error. In this case I want to return a wrapped ENOTCONN similar to what ceph calls do when mount is valid but not mounted. Either I return errNotConnected from validate _or_ I return it from every function that calls validate. The former, what I currently have in the code, seems more DRY to me as the callers of validate just pass on whatever error was returned as opposed to repeating 'return ..., errNotConnected` and also allowing for more error values in the future.